File: app/src/main/java/com/example/telemedicine_lab5/signaling/SocketSignalingClient.kt
package com.example.telemedicine_lab5.signaling

import io.socket.client.IO
import io.socket.client.Socket
import org.json.JSONObject
import org.webrtc.IceCandidate
import java.net.URI

class SocketSignalingClient {
    private var socket: Socket? = null

    fun connect(
        serverUrl: String,
        roomId: String,
        onConnected: () -> Unit,
        onReady: () -> Unit,
        onError: (String) -> Unit,
        onMessage: (SignalingMessage) -> Unit,
    ) {
        disconnect()

        val normalizedUrl = normalizeServerUrl(serverUrl)

        val options = IO.Options.builder()
            .setReconnection(true)
            .setReconnectionAttempts(10)
            .setTimeout(10_000)
            .setForceNew(true)
            .setPath("/socket.io/")
            // Use default websocket + polling
            .setTransports(arrayOf("websocket", "polling"))
            .build()

        socket = IO.socket(URI.create(normalizedUrl), options).apply {
            on(Socket.EVENT_CONNECT) {
                emit("join", roomId)
                onConnected()
            }

            on("ready") {
                onReady()
            }

            on("full") { args ->
                val room = args.firstOrNull()?.toString() ?: ""
                onError("Room $room is full")
            }

            on(Socket.EVENT_CONNECT_ERROR) { args ->
                onError(args.firstOrNull()?.toString() ?: "Socket connection error")
            }

            on("error") { args ->
                onError(args.firstOrNull()?.toString() ?: "Socket error")
            }

            on(Socket.EVENT_DISCONNECT) { args ->
                val reason = args.firstOrNull()?.toString() ?: "unknown"
                onError("Socket disconnected: $reason")
            }

            on("message") { args ->
                val payload = args.firstOrNull()
                if (payload is JSONObject) {
                    val normalizedPayload = if (payload.has("payload") && payload.opt("payload") is JSONObject) {
                        payload.optJSONObject("payload") ?: payload
                    } else {
                        payload
                    }
                    onMessage(parseSignalingMessage(normalizedPayload))
                }
            }

            connect()
        }
    }

    private fun normalizeServerUrl(serverUrl: String): String {
        val trimmed = serverUrl.trim()
        return if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
            trimmed
        } else {
            "http://$trimmed"
        }
    }

    fun sendOffer(roomId: String, sdp: String) {
        sendPayload(roomId, JSONObject().apply {
            put("type", "offer")
            put("sdp", sdp)
        })
    }

    fun sendAnswer(roomId: String, sdp: String) {
        sendPayload(roomId, JSONObject().apply {
            put("type", "answer")
            put("sdp", sdp)
        })
    }

    fun sendCandidate(roomId: String, candidate: IceCandidate) {
        sendPayload(roomId, JSONObject().apply {
            put("type", "candidate")
            put("candidate", candidate.sdp)
            put("sdpMid", candidate.sdpMid)
            put("sdpMLineIndex", candidate.sdpMLineIndex)
        })
    }

    private fun sendPayload(roomId: String, payload: JSONObject) {
        socket?.emit(
            "message",
            JSONObject().apply {
                put("room", roomId)
                put("payload", payload)
            },
        )
    }

    fun disconnect() {
        socket?.off()
        socket?.disconnect()
        socket = null
    }
}

File: flutter_app/lib/src/services/socket_signaling_client.dart
import 'dart:async';

import 'package:socket_io_client/socket_io_client.dart' as io;

import '../models/signaling_message.dart';

class SocketSignalingClient {
  io.Socket? _socket;
  final _messagesController = StreamController<SignalingMessage>.broadcast();

  Stream<SignalingMessage> get messages => _messagesController.stream;

  Future<void> connect({
    required String serverUrl,
    required String roomId,
    required void Function() onConnected,
    required void Function() onReady,
    required void Function(String error) onError,
  }) async {
    disconnect();

    final normalizedUrl = _normalizeServerUrl(serverUrl);
    _socket = io.io(
      normalizedUrl,
      io.OptionBuilder()
          .setPath('/socket.io/')
          .setTransports(['websocket', 'polling'])
          .enableReconnection()
          .setReconnectionAttempts(10)
          .setTimeout(10000)
          .build(),
    );

    _socket!
      ..onConnect((_) {
        _socket?.emit('join', roomId);
        onConnected();
      })
      ..on('ready', (_) => onReady())
      ..on('full', (payload) {
        onError('Room ${payload?.toString() ?? roomId} is full');
      })
      ..onConnectError((e) => onError(e?.toString() ?? 'Socket connection error'))
      ..onError((e) => onError(e?.toString() ?? 'Socket error'))
      ..onDisconnect((reason) => onError('Socket disconnected: $reason'))
      ..on('message', (payload) {
        final normalized = _normalizePayload(payload);
        if (normalized != null) {
          _messagesController.add(SignalingMessage.fromMap(normalized));
        }
      });

    _socket?.connect();
  }

  void sendOffer(String roomId, String sdp) {
    _sendPayload(roomId, {'type': 'offer', 'sdp': sdp});
  }

  void sendAnswer(String roomId, String sdp) {
    _sendPayload(roomId, {'type': 'answer', 'sdp': sdp});
  }

  void sendCandidate(String roomId, String candidate, String? sdpMid, int? sdpMLineIndex) {
    _sendPayload(roomId, {
      'type': 'candidate',
      'candidate': candidate,
      'sdpMid': sdpMid,
      'sdpMLineIndex': sdpMLineIndex,
    });
  }

  void _sendPayload(String roomId, Map<String, dynamic> payload) {
    _socket?.emit('message', {'room': roomId, 'payload': payload});
  }

  Map<String, dynamic>? _normalizePayload(dynamic payload) {
    if (payload is Map) {
      final casted = Map<String, dynamic>.from(payload as Map);
      final nested = casted['payload'];
      if (nested is Map) {
        return Map<String, dynamic>.from(nested as Map);
      }
      return casted;
    }
    return null;
  }

  String _normalizeServerUrl(String serverUrl) {
    final trimmed = serverUrl.trim();
    if (trimmed.startsWith('http://') || trimmed.startsWith('https://')) {
      return trimmed;
    }
    return 'http://$trimmed';
  }

  void disconnect() {
    _socket?.clearListeners();
    _socket?.disconnect();
    _socket = null;
  }

  void dispose() {
    disconnect();
    _messagesController.close();
  }
}
